system/security/plugins/jwtSecurity/jwtCache.cpp

```diff
@@ -0,0 +1,394 @@
+{
+    return _addScopePerm(wuScopePermMap, scope, getSecAccessFlagValue(accessFlagName.c_str()), defaultFlag, asDeny);
+}
+
+SecAccessFlags JWTUserInfo::matchWUScopePerm(const std::string& scope, SecAccessFlags defaultFlag) const
+{
+    return _matchScopePerm(wuScopePermMap, scope, defaultFlag);
+}
+
+JWTUserInfo& JWTUserInfo::eraseWUScopePerm(const std::string& scope)
+{
+    return _eraseScopePerm(wuScopePermMap, scope);
+}
+
+JWTUserInfo& JWTUserInfo::clearWUScopePerm()
+{
+    return _clearScopePerm(wuScopePermMap);
+}
+
+unsigned int JWTUserInfo::countWUScopePerms() const
+{
+    return _countScopePerms(wuScopePermMap);
+}
+
+bool JWTUserInfo::hasWUScopePerms() const
+{
+    return !wuScopePermMap.empty();
+}
+
+//-----------------------------------------------------------------------------
+
+JWTUserInfo& JWTUserInfo::addFileScopePerm(const std::string& scope, SecAccessFlags accessFlag, SecAccessFlags defaultFlag, bool asDeny)
+{
+    return _addScopePerm(fileScopePermMap, scope, accessFlag, defaultFlag, asDeny);
+}
+
+JWTUserInfo& JWTUserInfo::addFileScopePerm(const std::string& scope, const std::string& accessFlagName, SecAccessFlags defaultFlag, bool asDeny)
+{
+    return _addScopePerm(fileScopePermMap, scope, getSecAccessFlagValue(accessFlagName.c_str()), defaultFlag, asDeny);
+}
+
+SecAccessFlags JWTUserInfo::matchFileScopePerm(const std::string& scope, SecAccessFlags defaultFlag) const
+{
+    return _matchScopePerm(fileScopePermMap, scope, defaultFlag);
+}
+
+JWTUserInfo& JWTUserInfo::eraseFileScopePerm(const std::string& scope)
+{
+    return _eraseScopePerm(fileScopePermMap, scope);
+}
+
+JWTUserInfo& JWTUserInfo::clearFileScopePerm()
+{
+    return _clearScopePerm(fileScopePermMap);
+}
+
+unsigned int JWTUserInfo::countFileScopePerms() const
+{
+    return _countScopePerms(fileScopePermMap);
+}
+
+bool JWTUserInfo::hasFileScopePerms() const
+{
+    return !fileScopePermMap.empty();
+}
+
+//-----------------------------------------------------------------------------
+// Private methods
+//-----------------------------------------------------------------------------
+
+// regex patterns for converting a filename glob pattern
+// to an example string
+static const std::regex _globRegex1("\\[![0-9].+?\\]\\*?");     // foo[!1] -> fooa
+static const std::regex _globRegex2("\\[[0-9].+?\\]\\*?");      // foo[1] -> foo1
+static const std::regex _globRegex3("\\[![A-Za-z].+?\\]\\*?");  // foo[!m]* -> foo1
+static const std::regex _globRegex4("\\[[A-Za-z].+?\\]\\*?");   // foo[m]* -> fooa
+static const std::regex _globRegex5("[*?]");                    // foo* -> fooa, foo? -> fooa
+
+std::string JWTUserInfo::_globToExample(const std::string& pattern) const
+{
+    // Apply all filename glob patterns in order
+    return std::regex_replace(std::regex_replace(std::regex_replace(std::regex_replace(std::regex_replace(pattern, _globRegex1, "a"), _globRegex2, "1"), _globRegex3, "1"), _globRegex4, "a"), _globRegex5, "a");;
```

Comment:
I think this was clearer as separate lines (I don't think it is any more efficient).  What I was suggesting was combining some of the patterns.  E.g.

static const std::regex _globRegex23("\\([[0-9]|[![A-Za-z]).+?\\]\\*?");      // foo[1|!m] -> foo1

because it may be more efficient to match and there will be fewer replacements.  However it is only worth doing if it has a noticeable impact.

Reply:
Got it.  I thought you knew something I didn't re: argument usage and temp variables.

I've combined those two patterns you called out.  I did not combine the next two, however, because order is important there and there is no guarantee that the order within the pattern is honored.